### PR TITLE
feat(native-filters): Time native filter

### PR DESCRIPTION
**Pull request by simcha90**
### SUMMARY
This PR is continue of @agatapst PR: https://github.com/apache/superset/pull/12946

Here description of ☝️ PR:
```
The aim of this PR was to create foundations for new native filters' feature: time filters.
User can choose between one more filter type, which is Time.

That's the basics for this feature. What should be added in next PRs:

- UI improvements (if we want to show modal in modal, how to display very long time "pills" in Filter Bar etc.)
- Improve Filter Config Modal validation - right now datasource is required, but in case of time filters it should not
- Implement additional filtering: i.e. inverse
- Might also take a look on performance - it take more time to show time filters result than select filters result
- Filter indicator logic needs to be updated
```

In addition this PR fix some edge cases.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by amitmiran137** [comment]
cc: @junlincc

**Review by villebro — APPROVED** [review]
The native filter changes look good, but I refer the review of time picker changes to @zhaoyongjie .

**Comment by zhaoyongjie** [comment]
control layout issue on the latest Firefox 
image

**Comment by zhaoyongjie** [comment]
cc @junlincc 
`the range filter` should set the correct endpoints

image

**Comment by zhaoyongjie** [comment]
1. set native filter in dashboard
image

2. apply this filter in dashboard. select `last day` in time filter (no such data in `cleaned sales data`), expected display `no data` in each chart in the dashboard.
image

3. not apply filter
image

File: superset-frontend/src/explore/components/controls/DateFilterControl/DateFilterControl.tsx [hunk 1]
```diff
@@ -168,13 +168,20 @@ const IconWrapper = styled.span`
 interface DateFilterLabelProps {
   name: string;
   onChange: (timeRange: string) => void;
+  onTimeRangeChange?: (textValue?: string, timeRange?: string) => void;
```
Comment (zhaoyongjie) [inline superset-frontend/src/explore/components/controls/DateFilterControl/DateFilterControl.tsx:171]: `onChange` and `onTimeRangeChange` seem to have the same function, can we merge these props?

File: superset-frontend/src/filters/components/Time/AntdTimeFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  const { defaultValue, currentValue } = formData;
+
+  const [value, setValue] = useState<string>(defaultValue ?? DEFAULT_VALUE);
+
+  const handleTimeRangeChange = (textValue: string, timeRange: string) => {
+    setExtraFormData({
+      // @ts-ignore
+      extraFormData: {
+        override_form_data: {
+          time_range: timeRange,
```
Comment (zhaoyongjie) [inline superset-frontend/src/filters/components/Time/AntdTimeFilter.tsx:44]: Logic error here

File: superset-frontend/src/explore/components/controls/DateFilterControl/DateFilterControl.tsx [hunk 3]
```diff
@@ -76,12 +76,12 @@ const fetchTimeRange = async (
   const endpoint = `/api/v1/time_range/?q=${query}`;
   try {
     const response = await SupersetClient.get({ endpoint });
-    const timeRangeString = buildTimeRangeString(
-      response?.json?.result?.since || '',
-      response?.json?.result?.until || '',
-    );
+    const { since = '', until = '' } = response?.json?.result || {};
+    const timeRangeString = buildTimeRangeString(since, until);
```
Comment (zhaoyongjie) [inline superset-frontend/src/explore/components/controls/DateFilterControl/DateFilterControl.tsx:80]: NEED to double check here.
`timeRangeString` is `Absolute time range`. 
eg:
if today is 2021-2-8, then `last day`, generate "2021-02-07 : 2021-02-08".
when the 2021-3-8, then `last day` generate "2021-03-07 : 2021-03-08"

**Review by zhaoyongjie — COMMENTED** [review]
need double check

**Comment by junlincc** [comment]
@simcha90 thanks for taking over the project! really appreciate you work

tested and the basic functionality is working, can not fully test until the filter is connected to the indicator. @zhaoyongjie we should either skip the time column select or pre-select dttm column for your user, or simply narrow it down to show date column only in select dropdown. 
Also there's noticeable layout issue of regular filter on the left bar. 
Priority fo next steps  - 

1. Layout - make sure user can change filter value of regular filter (reason: we need to keep this feature in minimal demo-able state)
Screen Shot 2021-02-08 at 7 22 23 AM
2. connect to filter indicator and magnifier
3. Improve Filter Config Modal validation, populate only date/datetime/time columns in select 
4. Performance 
5. Inverse 

https://user-images.githubusercontent.com/67837651/107239584-d9efea80-69dd-11eb-88b7-7445f9c0bb01.mov

**Comment by junlincc** [comment]
> need double check

@villebro @simcha90 @zhaoyongjie what is the scope of this PR? 
 i listed my priority in last comment, can we make incremental changes based on the order?

**Comment by junlincc** [comment]
we also lost the indicator hook that we Agata previously implemented for the regular filter type. @simcha90 could you please double check?

**Comment by simcha90** [comment]
@junlincc I think scope of this PR - is ability to set TimeFilter as native filter and also set default value for it. All other additional features should be implemented in other PRs, and can be done in different times after this PR will be merged as base PR for it.

**Comment by zhaoyongjie** [comment]
Hi, @simcha90, @ville, @junlincc 

The implementation of this PR still needs to be considered. Now the filter passed by the native filter to each chart is the `specific time`. For example, if you choose,` Last week`, it will transform to `2021-01-01 : 2021-01-02`. This applies to The chart filter will not change. Imagine a scenario, this will cause a problem. After you select `Last week`, set the Dashboard to refresh by `every minute`. Then the chart will not update the filter.

File: superset-frontend/src/filters/components/Time/index.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import controlPanel from './controlPanel';
+import transformProps from './transformProps';
+import thumbnail from './images/thumbnail.png';
+
+export default class TimeFilterPlugin extends ChartPlugin {
+  constructor() {
+    const metadata = new ChartMetadata({
+      name: t('Time range filter plugin'),
+      description: 'Custom time filter plugin',
+      isNativeFilter: true,
```
Comment (simcha90) [inline superset-frontend/src/filters/components/Time/index.ts:29]: need update to behaviour

File: superset-frontend/src/filters/components/Time/index.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import controlPanel from './controlPanel';
+import transformProps from './transformProps';
+import thumbnail from './images/thumbnail.png';
+
+export default class TimeFilterPlugin extends ChartPlugin {
+  constructor() {
+    const metadata = new ChartMetadata({
+      name: t('Time range filter plugin'),
+      description: 'Custom time filter plugin',
+      behaviors: [Behavior.NATIVE_FILTER],
```
Comment (simcha90) [inline superset-frontend/src/filters/components/Time/index.ts:29]: it can be also `CROSS_FILTER`

**Comment by simcha90** [comment]
1. I think it have some inconsistencies, like if I choose `Custom` default value it shows like this
image

2. If I choose `previous week` it shows `dates` in filter bar instead of `previous week`, but after reload it shows `previous week`

image

File: superset-frontend/src/filters/components/Time/AntdTimeFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  useEffect(() => {
+    setValue(defaultValue ?? DEFAULT_VALUE);
+  }, [defaultValue]);
+
+  return (
+    // @ts-ignore
+    <Styles width={width}>
+      <DateFilterControl
+        value={value}
+        name=""
```
Comment (zhaoyongjie) [inline superset-frontend/src/filters/components/Time/AntdTimeFilter.tsx:65]: one nit:
could we set a default name for `DateFilterControl`

**Review by zhaoyongjie — APPROVED** [review]
LGTM, thanks ville.

**Comment by junlincc** [comment]
LGTM!  ✅ 
will connect to filter indicators and unit test in follow up PRs.

**Review by junlincc — APPROVED** [review]

**Pull request merged**